rs/crypto/internal/crypto_lib/bls12_381/vetkd/tests/tests.rs

```diff
@@ -1,60 +1,341 @@
 
-    let mut node_info = Vec::with_capacity(nodes);
+            let k = setup
+                .transport_sk
+                .decrypt(&ek, &proto.derived_pk, &proto.did)
+                .expect("Decryption failed");
 
-    for node in 0..nodes {
-        let node_sk = poly.evaluate_at(&Scalar::from_node_index(node as u32));
-        let node_pk = G2Affine::from(G2Affine::generator() * &node_sk);
+            keys_recovered.push(k);
+        } else {
+            assert!(
+                rec_threshold < threshold,
+                "Recovery fails with insufficient quorum"
+            );
+        }
+    }
+
+    // Now check that each recovered vetkey is the same:
+    let vetkey = keys_recovered[0].clone();
+
+    for k in &keys_recovered {
+        assert_eq!(*k, vetkey);
+    }
+
+    // Check that the vetkey output is a valid BLS signature
+    let msg = G1Affine::augmented_hash(proto.derived_pk.point(), &proto.did);
 
-        let eks = EncryptedKeyShare::create(rng, &master_pk, &node_sk, &tpk, &derivation_path, did);
+    assert!(verify_bls_signature(
+        &vetkey,
+        proto.derived_pk.point(),
+        &msg
+    ));
 
-        assert!(eks.is_valid(&master_pk, &node_pk, &derivation_path, did, &tpk));
+    // Check that if we introduce incorrect shares then combine_all will fail
 
-        // check that EKS serialization round trips:
-        let eks_bytes = eks.serialize();
-        let eks2 = EncryptedKeyShare::deserialize(eks_bytes).unwrap();
-        assert_eq!(eks, eks2);
+    let other_did = rng.gen::<[u8; 24]>();
+    let node_info_wrong_did = proto.create_encrypted_key_shares(rng, Some(&other_did));
 
-        node_info.push((node as u32, node_pk, eks));
+    let node_eks_wrong_did = node_info_wrong_did
+        .iter()
+        .map(|(idx, _pk, eks)| (*idx, eks.clone()))
+        .collect::<Vec<_>>();
+
+    // With combine_all even if we provide sufficiently many valid shares
+    // if any one share is invalid then combination will fail
+    for rec_threshold in 2..nodes {
+        let mut shares = random_subset(rng, &node_eks, rec_threshold - 1);
+        shares.append(&mut random_subset(rng, &node_eks_wrong_did, 1));
+        shares.shuffle(rng);
+        assert!(proto.combine_all(&shares).is_err());
     }
 
-    let ek = EncryptedKey::combine(
-        &node_info,
-        threshold,
-        &master_pk,
-        &tpk,
-        &derivation_path,
-        did,
-    )
-    .unwrap();
+    // With combine_valid_shares OTOH we detect and reject the shares
+
+    for rec_threshold in threshold..nodes {
+        let mut shares = random_subset(rng, &node_info, rec_threshold);
+        shares.append(&mut random_subset(rng, &node_info_wrong_did, 4));
+        shares.shuffle(rng);
 
-    let _k = tsk.decrypt(&ek, &dpk, did).unwrap();
+        let combined = proto.combine_valid(&shares).expect("Combination worked");
 
-    let derived_key = tsk
-        .decrypt_and_hash(&ek, &dpk, did, 32, b"aes-256-gcm-siv")
```

Comment:
Here `decrypt_and_hash` is currently not tested any more. I believe we have some tests involving `decrypt_and_hash` in `ic-vetkd-utils`. Are the `ic-vetkd-utils` maybe anyway the better place to test this? Do I understand correctly that the `TransportSecretKey` is actually more like a test utility here? If so, we could even consider removing the duplication and using the one from `ic-vetkd-utils` here...but not fully sure.

Reply:
Indeed so, this `TransportSecretKey` will never be used in prod since the key should be created on the client side (eg using webcrypto). I think it existed more as a demo of what we'd want the protocol to do, since the original code predates (and was spun out to) `ic-vetkd-utils`. I think it makes sense to remove `decrypt_and_hash` entirely from here.

Probably the correct final disposition is that `TransportSecretKey` is moved to the test code

Reply:
Moved `TransportSecretKey` to the test code since we're not going to need it for prod

Reply:
Sounds good, thanks! Andrea also reminded me that we should not forget to ensure that the `ic-vetkd-utils` (which is where the `TransportSecretKey` for production lives) are sufficiently tested. I believe there are tests at the moment, but not fully sure how comprehensive they are, so I created CRP-2669 accordingly.